perl.h

```diff
@@ -737,110 +737,156 @@ as in
 
 Note that you can't return a value out of this construct and cannot use it as
 an operand to the comma operator, which limit its utility.  If you need
-those features, instead use C language C<static inline> functions.
-But those won't work where the types of the inputs can vary depending on where
-called from; you'd have to create a separate function for each combination of
-parameter types.
+those features, instead use C language C<static inline> functions, if possible.
 
-Subtle bugs can occur due to this being a macro with names evaluated in
-the context of the caller, unlike a function.  These bugs can be eliminated by
-treating this more like a function.  This means that you must only use
-variables inside it that are either formal parameters or are declared
-internally within the construct.  Thus
+When using this construct, you need to keep in mind that this is a macro and
+not a function, hence
 
 =over
 
 =item *
 
-B<All> references to outside-the-construct variables must be through formal
-parameters.
+It is expanded within the name space of its caller (or chain of callers).  Any
+of those callers could have defined a symbol that is used inside your macro to
+be something, almost but not quite, entirely unlike what you expected.  Nobody
+is going to S<C<#define else  abort()>>, for example, even though it is
+theoretically possible, so you can assume that C language keywords and numeric
+constants have the expected meaning.
+
+But any variable name you choose I<could> clash with something up the caller
+chain.  There have been cases where this didn't happen until years after the
+macro was written.
+
+So you need to choose names for every variable inside your macro that are
+unlikely to be used by anyone outside it.
+
+One possible technique is to include the name of the macro as part of each
+variable name.  Suppose you have a macro named C<Foo_Bar> that has two
+variables that would normally be named C<sv> and C<len>.  You could instead
+name them C<sv__foo_bar> and C<len__foo_bar>, respectively, and pretty much
+guarantee that those names won't ever clash with anything else.  In this
+example naming convention, the macro name is appended to the variable name with
+two underscores, which makes the variable name stand out more, and indicates
+that the macro name is a modifier, and not the important part.  It also
+lowercases the macro name to make the result more like a typical variable name;
+this assumes that there isn't another macro named C<foo_bar>.
```

Comment:
It's not an issue for C, but since we build with C++: "Each identifier that contains a double underscore __ or begins with an underscore followed by an uppercase letter is reserved to the implementation for any use."

Reply:
revised